provisioner/template.go

```diff
@@ -733,3 +734,49 @@ func sumQuantities(quantities ...string) (string, error) {
 func awsValidID(id string) string {
 	return strings.Replace(id, ":", "__", -1)
 }
+
+// divideInstanceResourcesInRatio takes in an instance type and a ratio and returns the result of applying
+// the ratio on the instance CPU and Memory sizes.
+// The ratio must be a positive number and less than 1.
+// The resourceType must be either 'cpu' or 'memory'.
+// The result is rounded down to the nearest integer, to avoid overcommitting resources.
+// The units for CPU and memory are millicores and mebibytes (MiB) respectively.
+func divideInstanceResourcesInRatio(adapter *awsAdapter, instanceType string, resourceType string, ratio float32) (dividedResource int, err error) {
+	if adapter == nil || adapter.ec2Client == nil {
+		return 0, fmt.Errorf("the ec2 client is not available")
+	}
+
+	// validate the ratio
+	if ratio <= 0 {
+		return 0, fmt.Errorf("ratio must be a positive number")
+	} else if ratio > 1 {
+		return 0, fmt.Errorf("ratio must be less than 1")
+	} else if ratio == 1 {
+		return 0, fmt.Errorf("ratio cannot be 1, cannot use all resources of an instance")
+	}
+
+	// validate the resourceType
+	if resourceType != "cpu" && resourceType != "memory" {
+		return 0, fmt.Errorf("resourceType must be either 'cpu' or 'memory'")
+	}
+
+	// get the instance type info
+	input := ec2.DescribeInstanceTypesInput{InstanceTypes: awsUtil.StringSlice([]string{instanceType})}
+	output, err := adapter.ec2Client.DescribeInstanceTypes(&input)
```

Comment:
How often instance type configurations change? I'd assume its pretty-much constant so maybe, to avoid AWS API call on each template rendering its better to hardcode them or load all types once on CLM startup.

Reply:
I'm sure we have an in-memory representation of instance-types already in CLM, we use that in NodePool configuration iirc.

Reply:
You mean something like this here?
https://github.com/zalando-incubator/cluster-lifecycle-manager/blob/master/provisioner/node_pools.go#L308

Reply:
I see we populate the `InstanceTypes` data in `clusterpy.go` [here](https://github.com/zalando-incubator/cluster-lifecycle-manager/blob/master/provisioner/clusterpy.go#L313) and then pass it on to the new `caNodePoolProvisioner` struct [here](https://github.com/zalando-incubator/cluster-lifecycle-manager/blob/master/provisioner/clusterpy.go#L319). I'll figure out a way to use this in-memory representation in `template.go` functions as well. Will have to refactor some parts probably.

Reply:
I attempted one way I found of making this happen in the commit I pushed [here](https://github.com/zalando-incubator/cluster-lifecycle-manager/pull/743/commits/ea342af2f9ce76876183c73d5535a1b8ec3d5093).

Reply:
Just finished all the refactoring and updated the commit.